### DISCUSSION TITLE
Mixed-Integer

### DISCUSSION THREAD

**Discussion by malin33** [alphaville/optimization-engine#320] [Q&A] [answered]
Hi! 

Kindly asking for some advise or tips on how to implement a mixed-integer problem in OpEn. 
The approach I'm trying is to make the decision variable u an integer with the help of the constraint FiniteSet by: 
  
>   set=[[1]*N,[2]*N,[3]*N,[4]*N,[5]*N] 

> bounds=og.constraints.FiniteSet(set)

where N is the length of the prediction.
The problem then is that the solution only consists of a sequence with the same decision variable (for example [1,1,1,1,...]), and does not toggle between values (for example [1,1,1,1,2,2,2,2,...]) which is the behavior I'm looking for. 

My question is if the FiniteSet method is the way to go to make the decision variable u an integer or is there another way? 

Kind regards 
/Marcus

**Comment by alphaville** [comment, marked as answer]
Hi Marcus. If your decision variable is $(u_0, u_1, ..., u_{N-1})$, then you can use a Cartesian product of Finite sets. You can find an example [here](https://alphaville.github.io/optimization-engine/docs/python-interface#cartesian-product).

**Reply by malin33** [reply]
Hi, 
Thanks for the reply, I think I've got it to work using your suggestion! 
Including a example if anyone have a similar problem, probably not the best implementation but might be a start  

```
finiteSet=og.constraints.FiniteSet([[2],[3],[4],[5]]) 
#expand to horizon N 
sets=[finiteSet]*self.N
#Cartesan product to get combinations of finite sets
finalSetsConstraints=og.constraints.CartesianProduct(range(0,len(sets)),sets)
```
